### ISSUE TITLE
PRC File Versions and use of Schema

### ISSUE THREAD

**Issue by mvrhel** [pdf-association/pdf-issues#409]
** Where do you find updated file version information for PRC format??? **
I have run into many PRC files that have a version number of 15083.  Which indicates a version of the format that is from the 83rd day of the year 2015.  This is after the published spec.  Where do I get these revised versions?  The spec has a version number that is 10001 which means the 1st day of the year 2010.  So anyone trying to understand the spec is stuck with a version that is over 5 years  old.  I realize the schemas are meant to address this versioning, but the misprints and poor writing of the ISO published spec make it difficult.

**Comment by petervwyatt** [comment]
3D PDF agree that an informative note should be added concerning how files with undocumented versions should be considered. Note to be drafted by @stuartgalt .

**Comment by t3dr0** [comment]
We've requested this for years, in previous ISO meetings :)

**Comment by stuartgalt** [comment]
Any help in drafting the note would be greatly appreciated.

**Comment by mvrhel** [comment]
No, I am not seeing a minimal version for read beyond 10001.  I am following the rules you outlined above as that seems pretty clear in the spec.  My frustration is that someone is authoring versions that have no meaning to anyone other than the author.  For example, I generated a simple file from Solidworks.  It has a minimum read of 7094 and an author version 20225.  Wow.  This is the file that also has a type defined in a schema and then a schema defined for that type!   That just seems crazy to me.

Who ever created the particular file knows what is there and I suspect is not going to use the schema during reading but will probably rely upon the versioning number.  Everyone else has to make their way through the schema and hope that the schema encodings were done correctly -- or we can't read the file (but they can).  To me there seems to be something broken with the versioning in this approach when anyone can make up their own versions.  There is a user data element in many of the objects that is there to allow the storage of private data.  I really don't think the spec needs both that AND this wild west of versioning.

**Comment by mvrhel** [comment]
One idea is to deprecate the use of schemas and only documents with known published versions are considered to be conforming documents.  This avoids the issues I describe above.

**Comment by stuartgalt** [comment]
I agree that if the schema is used to add a new structure that is presented... that breaks interoperability.  But if the schema is used to support round tripping or being able to map the structure to external data of some sort there may be value.  Yes the private data is not useful to others but ideally the presentation would still work even if the additional information described in the schema is opaque?

Is there a middle ground?

**Comment by mvrhel** [comment]
There are plenty of places in the spec where user data can be stored, and the spec describes well the process of skipping it or knowing if that particular data was created by your writer so that your reader can deal with it.   This seems straightforward and really allows anyone to put whatever they want -- including an entire PRC stream if they so choose.  And it is simple for another reader to skip this chunk of data.   The schema system, which is tied to the version number places a pretty heavy burden on the reader and the writer.  The writer of the file could end up creating something that THEIR reader reads just fine because they know what data lurks in the bit stream, but perhaps they make a mistake in the schema specification, which other readers have to rely upon to navigate the stream -- so their readers fail.  If Adobe happens to be using such a reader that knows how the data is encoded due to the version number, then Acrobat renders the content fine, but another solution may not -- and may not render at all due to an error in the schema (files like this could easily be created).  This creates a competitive advantage to those readers that know the bit stream.  I would claim that this approach makes the specification fragile.  Especially since the decision of when a reader should or should not make use the schema is tied to the version number.  A number that should be related to an ISO specification that anyone can get.

**Comment by t3dr0** [comment]
The current ISO specification describes a single version. If any other version can be described, a new dated revision can be created. for STEP and GLTF annotations we've had two or more independent implementations before publishing. We're currently far from that for the use of schemas and newer versions.

**Comment by t3dr0** [comment]
Here's a new change proposal to be discussed. I have not really changed my mind about the above versioning draft, but is this new  edit a reasonable middle ground?

1. Add an additional NOTE statement in the opening section 4.2 Versioning after the second paragraph:
"This international standard specifies the PRC version10001. Be aware that documents with other PRC versions may alter the behaviour and presentation of content. Other versions as well as the logic and behaviour of PRC file schema are not addressed in this specification.

2. Edit section 8.1 - A PRC file reader
"Otherwise, the reader is built on a version of the specification that knows the structure of the base entities in the file and **will** use the schema definitions to process new data from the PRC file."
change to word "**will**" to "**shall**".

**Comment by mvrhel** [comment]
I still like the idea that other versions are considered to be non-compliant. Soon I am going to start looking at these files I have generated from solid works and I am dreading it due to the crazy schema use I mentioned above.  I don't know if they are using this to describe content that should be rendered but if so that should not be allowed unless they somehow publish the information. Allowing such a use creates an environment where a company that already has most of the market can keep everyone else out and really reduces the usefulness of the ISO standard.

**Comment by mvrhel** [comment]
This sounds reasonable to me. 

I am not privy to who is generating other versions other than what I have seen -- I am coming in rather late into this process.  As someone who is viewing this from the outside, it really should not matter.  What should matter is that the specification is clear what other unpublished versions mean. If we want a standard that is going to ensure the expected output by a reader that was implemented using the published standard, then allowing other versions without them being published leaves you with a broken standard -- and uncertainty if the output is as expected.  Not a good thing if you are relying upon this for critical part descriptions.

**Comment by petervwyatt** [comment]
Implemented

**Comment by t3dr0** [comment]
In follow-up after Draft #3 acceptance, here is an overview of this change.

1. This spec documents one version only.
2. Other versions may be created, marked by the author-version field. New versions can add new entities, behaviours or modified visual representations. Such new versions may claim compliance to PRC, without documentation to support the new changes.
3. The Schema system allows readers that don't know about the new version to "skip-over" the new material, i.e. to ignore it. However, without new version documentation there is no way to verify the skip-over is correct.
4. The Schema system is a complete software language, where code is interpreted at runtime. The code content of a Schema can do much more than simply skip-over. In theory it could expose security vulnerabilities.

Therefore, this change marks any undocumented versions non-compliant with the PRC specification. If such documentation becomes available, dated revisions could be created to cover them.

Effects:
This change allows implementers to follow this spec and verify correct behaviour and reduces vulnerability exposure.
However it also means many existing PRCs in the field (using different versions numbers) become non-compliant.
